onboarding/fetch-bbluechip (FetchWithStrapiRequest)

bbluechip opened the pull request:
Changed ResponsiveCard design.

7alip commented on packages/ui/src/examples/E0_ResponsiveCard.tsx:
```diff
@@ -1,33 +1,63 @@
-  const randomPic = 'https://picsum.photos/600/400'
+  const randomPic =
+    "https://www.formica.com/-/media/project/formica/emea/products/swatch-images/f2828/f2828-swatch.jpg";
+  const isWideVersion = useBreakpointValue({ base: false, md: true });
```
`useBreakpointValue` uses JavaScript to calculate the viewport width under the hood. We focus on using pure css approach when making our components responsive.

For example, in plain css we use [media queries.](https://www.w3schools.com/css/css3_mediaqueries.asp). ChakraUI and some other competitors try to simplify using media queries.

**With ChakraUI**
```tsx
<Box width={{ base: "full", md: "50%", lg: "25%" }} /> // Object notation
// or
<Box width={["full", "full", "50%", "25%"] /> // Array notation
```

**With TailwindCSS**
```tsx
<div className="w-full md:w-1/2 lg:w-1/4"  />
```

**With plain css**
```css
div {
  width: 100%;
}

@media screen and (min-width: 30em) {
  div {
    width: 50%;
  }
}

@media screen and (min-width: 48em) {
  div {
    width: 25%;
  }
}
```

> Remember that the **Mobile First** approach uses `(min-width)` in the queries, while the **Desktop First** approach uses `(max-width)`. And most library uses mobile first approach.

7alip reviewed (commented):
Thanks. I left a comment about your solution.

7alip commented on packages/ui/src/examples/E1_FetchWithAxios.tsx:
```diff
@@ -2,24 +2,45 @@
+const BLOG_URL = "https://wsvv-api-staging.onrender.com/api/blogs?locale=tr"
 
 export const FetcWithAxios = () => {
   const [blogs, setBlogs] = useState([])
```
```suggestion
  const [blogs, setBlogs] = useState<Blog[]>([])
```

You can specify the type of the state. It helps.

7alip commented on packages/ui/src/examples/E1_FetchWithAxios.tsx:
```diff
@@ -2,24 +2,45 @@
+  }, [locale])
 
-  return <Box>{/* TODO: Show only title of the blogs */}</Box>
+  return
+   (
+  <Box>
+    {blogs.map((blog, index) => (
+    <div key={index}>{blog}</div>
+  ))}
+  <p>hello</p>
+  </Box>
+  )
```
It seems your code is not formatted. You can check if you have [EsLint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) installed in VSCode. I recommend using VSCode. If you have any other IDE, you can investigate how to format your code based on the [prettier config](https://github.com/wsvvrijheid/monorepo/blob/main/.prettierrc) in the root of the project.

7alip reviewed (approved):
Thanks

7alip commented on packages/ui/src/examples/E1_FetchWithAxios.tsx:
```diff
@@ -2,24 +2,45 @@
+   (
+  <Box>
+    {blogs.map((blog, index) => (
+    <div key={index}>{blog}</div>
```
`{blog}` is not renderable by React because it's an object. 

```suggestion
    <div key={index}>{blog.title}</div>
```

7alip commented on packages/ui/src/examples/E1_FetchWithAxios.tsx:
```diff
@@ -2,24 +2,45 @@
+  }, [locale])
 
-  return <Box>{/* TODO: Show only title of the blogs */}</Box>
+  return
```
@bbluechip You won't see the results in Storybook, the reason is that you don't return the components but undefined because of unformatted code.

Please right click on the code and select `Format Document`. Make sure you have Prettier and ESlint extensions. When you format your code, you will notice that the JSX part of your component is unreachable.

| Original (unformatted)                                                                                 | Formatted                                                                                              | Correct                                                                                                |
| ------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------ |
| ![image](https://github.com/wsvvrijheid/monorepo/assets/22167684/308c2c48-0be4-46f8-bef3-6863a5fc5f9a) | ![image](https://github.com/wsvvrijheid/monorepo/assets/22167684/407143a9-0e26-449e-a0ad-4af32798de5b) | ![image](https://github.com/wsvvrijheid/monorepo/assets/22167684/683c9eac-0775-4649-8dca-550df9fa23c3) |

7alip replied:
You can also see that the JSX has pale colors which means this part is not reachable because of the early return.

<img width="1286" alt="Screenshot 2024-01-04 at 03 24 19" src="https://github.com/wsvvrijheid/monorepo/assets/22167684/3e636e9c-50fa-4f0e-a60a-265dbb7f43fa">

7alip commented on packages/ui/src/examples/E1_FetchWithAxios.tsx:
```diff
@@ -1,25 +1,33 @@
+    <Box>
+      <ul>
+        {blogs.map((blog) => (
+          <li key={blog.id}>{blog.title}</li>
```
TypeScript probably complains about `blog.id` and `blog.title`, because blog type is not specified. You can indicate the blog state type as following:

```tsx
const [blogs, setBlogs] = useState<Blog[]>([])
```

bbluechip commented:
Added the interface of the Blog for not returning an error in useState line, need to specify but not defining a variable. 

Added response.data.data => First I thought respose.data would be enough but then I remembered, response.data is just the base returning data. Because of the api has data, meta objects. I need to write one more .data after response.data 

In this version FetchWithAxios running well.

7alip reviewed (approved):
Thanks. You can close the task now :)

7alip commented on packages/ui/src/examples/E2_FetchWithStrapiRequest.tsx:
```diff
@@ -1,17 +1,51 @@
+    const fetchData = async () => {
+      try {
+        // I couldn't find any argument for api url in strapiRequest function
+        const response = await strapiRequest({
```
As the name suggests, `strapiRequest` is simply an axios wrapper for fetching data from our Strapi backend. It's a generic function (`function<T> {}`) that allows you to specify which type of Strapi model _(Blog, Activity, Hashtag etch)_ you want to retrieve data from.

```suggestion
        const response = await strapiRequest<Blog>({
```

7alip commented on packages/ui/src/examples/E2_FetchWithStrapiRequest.tsx:
```diff
@@ -1,17 +1,51 @@
+        // I couldn't find any argument for api url in strapiRequest function
+        const response = await strapiRequest({
+          endpoint: 'blogs',
+          token: PUBLIC_TOKEN,
```
You should specify the `locale` for some models, otherwise only the models in the default language (en) will be retrieved.

```suggestion
          locale: 'tr' // or preferably locale from useRouter()
          token: PUBLIC_TOKEN,
```

7alip commented on packages/ui/src/examples/E2_FetchWithStrapiRequest.tsx:
```diff
@@ -1,17 +1,51 @@
-  return <Box>{/* TODO: Show only title of the blogs */}</Box>
-}
+        // property data does not exist on type 'StrapiModel[] so I write an if statement here 
+        if ('data' in response.data)
+          setBlogs(response.data.data as Blog[])
```
This statement is not necessary because `strapiRequest` should return as the following format: 
`{ data: [], pagination: {} }`


There are some other advanced debugging options, such as Network tab.

You can check to which endpoint are you making a request, what is your payload and what the server responses.

In your case, you are probably making a request to `http://localhost:1337`, which might not be running locally.

**You have some options to make it work:**

1. Update your env  `NEXT_PUBLIC_API_URL=http://localhost:1337` to`NEXT_PUBLIC_API_URL=https://wsvv-api-staging.onrender.com` in **packages/ui/.env**
2. Run your local server  by adding `"storybook": "strapi dev"` to your `apps/api/package.json` scripts.
3. Run your local server by running `cd apps/api && yarn dev`
4. If you are unable fix with these options, you can statically change API_URL in your strapiRequest function

<img width="1826" alt="Screenshot 2024-01-05 at 22 31 31" src="https://github.com/wsvvrijheid/monorepo/assets/22167684/d02cba63-539c-43e7-b739-21d16f25607c">

7alip reviewed (commented):
Thanks, I think you should just update your env variables or run your local server to make it work.
